x/expiration/keeper/keeper.go

```diff
@@ -0,0 +1,565 @@
+	if err := k.validateSigners(ctx, expiration.Owner, signers, msgTypeURL); err != nil {
+		return expiration, types.ErrInvalidSigners.Wrap(err.Error())
+	}
+
+	return expiration, nil
+}
+
+// ResolveDepositor resolves the account address where the expiration message deposit will be refunded.
+func (k Keeper) ResolveDepositor(
+	ctx sdk.Context,
+	expiration types.Expiration,
+	msg *types.MsgInvokeExpirationRequest,
+) (sdk.AccAddress, error) {
+	// check for owner in list first
+	for _, signer := range msg.Signers {
+		if signer == expiration.Owner {
+			addr, err := sdk.AccAddressFromBech32(expiration.Owner)
+			if err != nil {
+				return nil, types.ErrResolveDepositor.Wrap(
+					types.ErrInvalidSigners.Wrapf("[%s]: %v", signer, err).Error())
+			}
+			return addr, nil
+		}
+	}
+	// fall back to first signer if after expiration
+	if ctx.BlockTime().After(expiration.Time) && len(msg.Signers) >= 1 {
+		signer := msg.Signers[0]
+		addr, err := sdk.AccAddressFromBech32(signer)
+		if err != nil {
+			return nil, types.ErrResolveDepositor.Wrap(
+				types.ErrInvalidSigners.Wrapf("[%s]: %v", signer, err).Error())
+		}
+		return addr, nil
+	}
+
+	// error if no qualifying depositors are found
+	return nil, types.ErrResolveDepositor.Wrapf("asset [%s]", msg.ModuleAssetId)
+}
+
+// Private method that validates expiration message signatures
+func (k Keeper) validateSigners(
+	ctx sdk.Context,
+	owner string,
+	signers []string,
+	msgTypeURL string,
+) error {
+	found := false
+	for _, signer := range signers {
+		if signer == owner {
+			found = true
+			break
+		}
+
+		// validate signer with authz
+		var err error
+		found, err = k.hasSignerWithAuthz(ctx, owner, signer, msgTypeURL)
+		if err != nil {
+			return err
+		}
+		if found {
+			break
+		}
+	}
+
+	if !found {
+		return fmt.Errorf("intended signers %s do not match given signer [%s]", signers, owner)
+	}
+
+	return nil
+}
+
+// Private method that checks if any of the signers have been
+// granted authorizations to perform actions on an expiration record
+func (k Keeper) hasSignerWithAuthz(
+	ctx sdk.Context,
+	owner string,
+	signer string,
+	msgTypeURL string,
+) (bool, error) {
+	granter, err := sdk.AccAddressFromBech32(owner)
+	if err != nil {
+		return false, fmt.Errorf("invalid owner: %w", err)
+	}
+	grantee, err := sdk.AccAddressFromBech32(signer)
+	if err != nil {
+		return false, fmt.Errorf("invalid signers: %w", err)
+	}
+
+	authorization, exp := k.authzKeeper.GetAuthorization(ctx, grantee, granter, msgTypeURL)
+	if authorization != nil {
+		resp, err := authorization.Accept(ctx, nil)
+		if err != nil {
+			return false, err
+		}
+		if resp.Accept {
+			switch {
+			case resp.Delete:
+				err = k.authzKeeper.DeleteGrant(ctx, grantee, granter, msgTypeURL)
+				if err != nil {
+					return false, err
+				}
+			case resp.Updated != nil:
+				if err = k.authzKeeper.SaveGrant(ctx, grantee, granter, resp.Updated, exp); err != nil {
+					return false, err
+				}
```

Comment:
This is going to require some discussion.

1. Do the signers on an expiration's message need to be validated at time of expiration creation?
2. When it gets invoked?
3. When should any `authz` authorization resources be consumed? At the very least, I know they need to only be consumed once.
4. Should they be validated at all?

If we remove the `AddExpiration` endpoint, and only allow creation of expirations via code in other modules, I feel like we shouldn't care about the signers at all. Simply trust that the module isn't trying to do anything nefarious.

If we don't check the signers when the expiration is invoked, it can lead to a situation where a thing is created by one person (along with its expiration), then sent (sold?) to another (i.e. the things owner changes), then the original owner triggers the expiration of the thing. If expirations can only be created by another module, it'd be up to that module to properly update the expiration entry to reflect the new owner.

If we do check the signers when the expiration is invoked, what should happen when that validation fails? Again, this could happen when a things owners change, but the expiration is updated. We could easily end up stuck with an expiration entry that can't be invoked by anyone.

If `authz` gets involved in that signer checking, things get really crazy. It'd be easy for an authorization to expire between when an expiration is created and when it ends. Basically, there's no guarantee that an authorization that is valid now, will still be valid (or applicable) in the future. Conversely, just because an authorization doesn't exist now, doesn't mean it won't in the future. As such, I feel like there's no valid answer to "When should an authorization's resourced be used?"